core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -127,17 +131,111 @@ KOKKOS_INLINE_FUNCTION
     std::enable_if_t<std::is_same<T, unsigned long long>::value, T>
         cast_from_half(half_t);
 
-class half_t {
+class alignas(2) half_t {
  public:
   using impl_type = Kokkos::Impl::half_impl_t::type;
 
  private:
   impl_type val;
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void impl_volatile_add(const volatile half_t src) const volatile {
+    // Kokkos::atomic_store and atomicAdd not available on cuda < 10
+#if (CUDA_VERSION < 10000)
+    alignas(4) volatile float lhs;
+    alignas(4) volatile float rhs;
+    lhs                         = __half2float(const_cast<impl_type&>(val));
+    rhs                         = __half2float(const_cast<impl_type&>(src.val));
```

Comment:
As noted elsewhere, this is negating the `volatile` semantics of the signature, and possibly UB. Load from the `volatile` member into a temporary, operate on that as needed, and then (if possible, since it's implied that it's not) store back to the member without casting away `volatile`.

Reply:
If, as an earlier comment suggests, CUDA doesn't actually support loads from `volatile half_t`, then do it with a `volatile` `memcpy`-like operation on `uint16_t` or the two underlying bytes